File: strategies/3-greedy/FractionalKnapsack.java
import java.util.*;

/**
 * Greedy Strategy: Fractional Knapsack Problem
 * Core Idea: Always choose item with highest value-to-weight ratio first
 * Time Complexity: O(n log n) due to sorting
 * Space Complexity: O(n) for storing items and results
 */
public class FractionalKnapsack {
    
    /**
     * Item class to represent knapsack items
     */
    static class Item {
        int weight;
        int value;
        double valuePerWeight;
        String name;
        
        Item(String name, int weight, int value) {
            this.name = name;
            this.weight = weight;
            this.value = value;
            this.valuePerWeight = (double) value / weight;
        }
        
        @Override
        public String toString() {
            return String.format("%s (w:%d, v:%d, v/w:%.2f)", 
                               name, weight, value, valuePerWeight);
        }
    }
    
    /**
     * Result class to store knapsack solution
     */
    static class KnapsackResult {
        List<Item> selectedItems;
        List<Double> fractions;
        double totalValue;
        int totalWeight;
        
        KnapsackResult() {
            selectedItems = new ArrayList<>();
            fractions = new ArrayList<>();
            totalValue = 0;
            totalWeight = 0;
        }
        
        void addItem(Item item, double fraction) {
            selectedItems.add(item);
            fractions.add(fraction);
            totalValue += item.value * fraction;
            totalWeight += item.weight * fraction;
        }
    }
    
    /**
     * Solve fractional knapsack using greedy approach
     * @param items Array of available items
     * @param capacity Maximum weight capacity of knapsack
     * @return KnapsackResult containing optimal solution
     */
    public static KnapsackResult solveKnapsack(Item[] items, int capacity) {
        KnapsackResult result = new KnapsackResult();
        
        // Step 1: Sort items by value-to-weight ratio in descending order (greedy choice)
        Arrays.sort(items, (a, b) -> Double.compare(b.valuePerWeight, a.valuePerWeight));
        
        int remainingCapacity = capacity;
        
        // Step 2: Process items in order of decreasing value-to-weight ratio
        for (Item item : items) {
            if (remainingCapacity == 0) {
                break; // Knapsack is full
            }
            
            // Greedy choice: take as much of current item as possible
            if (item.weight <= remainingCapacity) {
                // Take entire item
                result.addItem(item, 1.0);
                remainingCapacity -= item.weight;
            } else {
                // Take fractional part of item
                double fraction = (double) remainingCapacity / item.weight;
                result.addItem(item, fraction);
                remainingCapacity = 0; // Knapsack is now full
            }
        }
        
        return result;
    }
    
    /**
     * Print detailed solution
     */
    public static void printSolution(KnapsackResult result, int capacity) {
        System.out.println("Optimal solution:");
        System.out.println("Items selected (in order of selection):");
        
        for (int i = 0; i < result.selectedItems.size(); i++) {
            Item item = result.selectedItems.get(i);
            double fraction = result.fractions.get(i);
            
            if (fraction == 1.0) {
                System.out.printf("  %s - Full item (fraction: %.0f%%)\n", 
                                item, fraction * 100);
            } else {
                System.out.printf("  %s - Partial item (fraction: %.1f%%)\n", 
                                item, fraction * 100);
            }
        }
        
        System.out.printf("Total value: %.2f\n", result.totalValue);
        System.out.printf("Total weight: %.1f / %d\n", result.totalWeight, capacity);
        System.out.printf("Efficiency: %.2f%% capacity used\n", 
                         (result.totalWeight / capacity) * 100);
    }
    
    /**
     * Print all available items sorted by value-to-weight ratio
     */
    public static void printItemsByRatio(Item[] items) {
        // Create copy to avoid modifying original array
        Item[] sortedItems = Arrays.copyOf(items, items.length);
        Arrays.sort(sortedItems, (a, b) -> Double.compare(b.valuePerWeight, a.valuePerWeight));
        
        System.out.println("Items sorted by value-to-weight ratio (greedy order):");
        for (int i = 0; i < sortedItems.length; i++) {
            System.out.printf("  %d. %s\n", i + 1, sortedItems[i]);
        }
    }
    
    /**
     * Compare with 0/1 knapsack approach (taking whole items only)
     */
    public static void compareWith01Knapsack(Item[] items, int capacity) {
        // Sort by value-to-weight ratio
        Item[] sortedItems = Arrays.copyOf(items, items.length);
        Arrays.sort(sortedItems, (a, b) -> Double.compare(b.valuePerWeight, a.valuePerWeight));
        
        int totalValue = 0;
        int totalWeight = 0;
        List<Item> selectedItems = new ArrayList<>();
        
        for (Item item : sortedItems) {
            if (totalWeight + item.weight <= capacity) {
                selectedItems.add(item);
                totalValue += item.value;
                totalWeight += item.weight;
            }
        }
        
        System.out.println("0/1 Knapsack comparison (whole items only):");
        System.out.println("Selected items: " + selectedItems.size());
        for (Item item : selectedItems) {
            System.out.println("  " + item);
        }
        System.out.println("Total value: " + totalValue);
        System.out.println("Total weight: " + totalWeight + " / " + capacity);
    }
    
    public static void main(String[] args) {
        System.out.println("=== Fractional Knapsack - Greedy Algorithm ===");
        
        // Test Case 1: Classic example
        System.out.println("Test Case 1: Classic fractional knapsack");
        Item[] items1 = {
            new Item("Gold", 10, 60),
            new Item("Silver", 20, 100),
            new Item("Bronze", 30, 120)
        };
        int capacity1 = 50;
        
        System.out.println("Available items:");
        for (Item item : items1) {
            System.out.println("  " + item);
        }
        System.out.println("Knapsack capacity: " + capacity1);
        System.out.println();
        
        printItemsByRatio(items1);
        System.out.println();
        
        KnapsackResult result1 = solveKnapsack(items1, capacity1);
        printSolution(result1, capacity1);
        System.out.println();
        
        compareWith01Knapsack(items1, capacity1);
        System.out.println();
        
        // Test Case 2: All items fit exactly
        System.out.println("Test Case 2: All items fit exactly");
        Item[] items2 = {
            new Item("Item A", 5, 10),
            new Item("Item B", 4, 40),
            new Item("Item C", 6, 30),
            new Item("Item D", 3, 50)
        };
        int capacity2 = 18; // Exactly sum of all weights
        
        System.out.println("Available items:");
        for (Item item : items2) {
            System.out.println("  " + item);
        }
        System.out.println("Knapsack capacity: " + capacity2);
        System.out.println();
        
        KnapsackResult result2 = solveKnapsack(items2, capacity2);
        printSolution(result2, capacity2);
        System.out.println();
        
        // Test Case 3: Very small capacity
        System.out.println("Test Case 3: Very small capacity");
        Item[] items3 = {
            new Item("Diamond", 1, 15),
            new Item("Ruby", 2, 20),
            new Item("Emerald", 3, 25)
        };
        int capacity3 = 2;
        
        System.out.println("Available items:");
        for (Item item : items3) {
            System.out.println("  " + item);
        }
        System.out.println("Knapsack capacity: " + capacity3);
        System.out.println();
        
        KnapsackResult result3 = solveKnapsack(items3, capacity3);
        printSolution(result3, capacity3);
        System.out.println();
        
        // Test Case 4: Single item larger than capacity
        System.out.println("Test Case 4: Single item larger than capacity");
        Item[] items4 = {
            new Item("Heavy Gold", 15, 100)
        };
        int capacity4 = 10;
        
        System.out.println("Available items:");
        for (Item item : items4) {
            System.out.println("  " + item);
        }
        System.out.println("Knapsack capacity: " + capacity4);
        System.out.println();
        
        KnapsackResult result4 = solveKnapsack(items4, capacity4);
        printSolution(result4, capacity4);
        System.out.println();
        
        // Test Case 5: Items with same value-to-weight ratio
        System.out.println("Test Case 5: Items with same value-to-weight ratio");
        Item[] items5 = {
            new Item("Item X", 2, 10), // ratio = 5.0
            new Item("Item Y", 4, 20), // ratio = 5.0
            new Item("Item Z", 6, 30)  // ratio = 5.0
        };
        int capacity5 = 8;
        
        System.out.println("Available items:");
        for (Item item : items5) {
            System.out.println("  " + item);
        }
        System.out.println("Knapsack capacity: " + capacity5);
        System.out.println();
        
        KnapsackResult result5 = solveKnapsack(items5, capacity5);
        printSolution(result5, capacity5);
        System.out.println("Note: When ratios are equal, algorithm processes in current order");
    }
}

File: strategies/3-greedy/FractionalKnapsack.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Greedy Strategy: Fractional Knapsack Problem
 * Core Idea: Always pick items with highest value-to-weight ratio first
 * Time Complexity: O(n log n) due to sorting
 * Space Complexity: O(1) excluding input
 */

/**
 * Item structure for the knapsack problem
 */
typedef struct {
    int value;
    int weight;
    double valueToWeightRatio;
    int index; // Original index for tracking
} Item;

/**
 * Compare function for sorting items by value-to-weight ratio (descending)
 */
int compareByRatio(const void* a, const void* b) {
    Item* itemA = (Item*)a;
    Item* itemB = (Item*)b;
    
    // Sort in descending order of value-to-weight ratio
    if (itemA->valueToWeightRatio < itemB->valueToWeightRatio) return 1;
    if (itemA->valueToWeightRatio > itemB->valueToWeightRatio) return -1;
    return 0;
}

/**
 * Calculate value-to-weight ratios for all items
 */
void calculateRatios(Item items[], int n) {
    for (int i = 0; i < n; i++) {
        if (items[i].weight > 0) {
            items[i].valueToWeightRatio = (double)items[i].value / items[i].weight;
        } else {
            items[i].valueToWeightRatio = 0.0; // Handle zero weight edge case
        }
    }
}

/**
 * Solve fractional knapsack using greedy approach
 * @param items Array of items
 * @param n Number of items
 * @param capacity Knapsack capacity
 * @param selected Array to store selection ratios (0.0 to 1.0)
 * @return Maximum value achievable
 */
double fractionalKnapsack(Item items[], int n, int capacity, double selected[]) {
    // Step 1: Calculate value-to-weight ratios
    calculateRatios(items, n);
    
    // Step 2: Sort items by value-to-weight ratio (greedy choice preparation)
    qsort(items, n, sizeof(Item), compareByRatio);
    
    // Initialize selection array
    for (int i = 0; i < n; i++) {
        selected[items[i].index] = 0.0;
    }
    
    double totalValue = 0.0;
    int remainingCapacity = capacity;
    
    // Step 3: Greedily select items
    for (int i = 0; i < n && remainingCapacity > 0; i++) {
        if (items[i].weight <= remainingCapacity) {
            // Take the whole item
            selected[items[i].index] = 1.0;
            totalValue += items[i].value;
            remainingCapacity -= items[i].weight;
        } else {
            // Take fraction of the item
            double fraction = (double)remainingCapacity / items[i].weight;
            selected[items[i].index] = fraction;
            totalValue += items[i].value * fraction;
            remainingCapacity = 0;
        }
    }
    
    return totalValue;
}

/**
 * Print items information
 */
void printItems(Item originalItems[], int n) {
    printf("Items (Value, Weight, Ratio):\n");
    for (int i = 0; i < n; i++) {
        double ratio = (originalItems[i].weight > 0) ? 
                      (double)originalItems[i].value / originalItems[i].weight : 0.0;
        printf("  Item %d: (%d, %d, %.2f)\n", 
               i, originalItems[i].value, originalItems[i].weight, ratio);
    }
}

/**
 * Print selection details
 */
void printSelection(Item originalItems[], int n, double selected[], 
                   double totalValue, int capacity) {
    printf("Selection (Item, Fraction, Value Obtained, Weight Used):\n");
    double totalWeight = 0.0;
    
    for (int i = 0; i < n; i++) {
        if (selected[i] > 0.0) {
            double valueObtained = originalItems[i].value * selected[i];
            double weightUsed = originalItems[i].weight * selected[i];
            totalWeight += weightUsed;
            
            printf("  Item %d: %.1f%%, Value: %.2f, Weight: %.2f\n", 
                   i, selected[i] * 100, valueObtained, weightUsed);
        }
    }
    
    printf("Total Value: %.2f\n", totalValue);
    printf("Total Weight Used: %.2f / %d\n", totalWeight, capacity);
}

/**
 * Demonstrate the greedy approach step by step
 */
void demonstrateGreedySteps(Item items[], int n, int capacity) {
    printf("Greedy Algorithm Steps:\n");
    
    // Calculate and display ratios
    calculateRatios(items, n);
    
    // Create a copy for sorting without affecting original
    Item* sortedItems = (Item*)malloc(n * sizeof(Item));
    for (int i = 0; i < n; i++) {
        sortedItems[i] = items[i];
    }
    
    qsort(sortedItems, n, sizeof(Item), compareByRatio);
    
    printf("1. Sort by value-to-weight ratio (descending):\n");
    for (int i = 0; i < n; i++) {
        printf("   Item %d: ratio %.2f\n", 
               sortedItems[i].index, sortedItems[i].valueToWeightRatio);
    }
    
    printf("2. Greedy selection:\n");
    int remainingCapacity = capacity;
    double totalValue = 0.0;
    
    for (int i = 0; i < n && remainingCapacity > 0; i++) {
        if (sortedItems[i].weight <= remainingCapacity) {
            printf("   Take full Item %d (weight %d, value %d) - Remaining capacity: %d\n",
                   sortedItems[i].index, sortedItems[i].weight, 
                   sortedItems[i].value, remainingCapacity - sortedItems[i].weight);
            totalValue += sortedItems[i].value;
            remainingCapacity -= sortedItems[i].weight;
        } else {
            double fraction = (double)remainingCapacity / sortedItems[i].weight;
            printf("   Take %.1f%% of Item %d (weight %.2f, value %.2f) - Capacity exhausted\n",
                   fraction * 100, sortedItems[i].index, 
                   sortedItems[i].weight * fraction, sortedItems[i].value * fraction);
            totalValue += sortedItems[i].value * fraction;
            remainingCapacity = 0;
        }
    }
    
    printf("3. Final total value: %.2f\n", totalValue);
    free(sortedItems);
}

int main() {
    printf("=== Fractional Knapsack - Greedy Algorithm ===\n");
    
    // Test Case 1: Classic example
    printf("Test Case 1: Classic fractional knapsack\n");
    Item items1[] = {
        {60, 10, 0.0, 0},   // Ratio: 6.0
        {100, 20, 0.0, 1},  // Ratio: 5.0
        {120, 30, 0.0, 2}   // Ratio: 4.0
    };
    int n1 = 3;
    int capacity1 = 50;
    double selected1[3];
    
    printItems(items1, n1);
    printf("Knapsack capacity: %d\n\n", capacity1);
    
    double maxValue1 = fractionalKnapsack(items1, n1, capacity1, selected1);
    printSelection(items1, n1, selected1, maxValue1, capacity1);
    printf("\n");
    
    // Demonstrate greedy approach
    Item items1Copy[] = {
        {60, 10, 0.0, 0},
        {100, 20, 0.0, 1},
        {120, 30, 0.0, 2}
    };
    demonstrateGreedySteps(items1Copy, n1, capacity1);
    printf("Expected: Take Item 0 (full), Item 1 (full), Item 2 (2/3)\n");
    printf("Optimal value: %.2f\n\n", maxValue1);
    
    // Test Case 2: All items fit
    printf("Test Case 2: All items fit in knapsack\n");
    Item items2[] = {
        {10, 5, 0.0, 0},    // Ratio: 2.0
        {40, 4, 0.0, 1},    // Ratio: 10.0
        {30, 6, 0.0, 2},    // Ratio: 5.0
        {50, 3, 0.0, 3}     // Ratio: 16.67
    };
    int n2 = 4;
    int capacity2 = 20;
    double selected2[4];
    
    printItems(items2, n2);
    printf("Knapsack capacity: %d\n\n", capacity2);
    
    double maxValue2 = fractionalKnapsack(items2, n2, capacity2, selected2);
    printSelection(items2, n2, selected2, maxValue2, capacity2);
    printf("Note: All items fit, so we take everything\n\n");
    
    // Test Case 3: Only one item fits
    printf("Test Case 3: Only one item fits completely\n");
    Item items3[] = {
        {20, 15, 0.0, 0},   // Ratio: 1.33
        {30, 10, 0.0, 1},   // Ratio: 3.0
        {40, 20, 0.0, 2}    // Ratio: 2.0
    };
    int n3 = 3;
    int capacity3 = 15;
    double selected3[3];
    
    printItems(items3, n3);
    printf("Knapsack capacity: %d\n\n", capacity3);
    
    double maxValue3 = fractionalKnapsack(items3, n3, capacity3, selected3);
    printSelection(items3, n3, selected3, maxValue3, capacity3);
    printf("Greedy choice: Take item with highest ratio first\n\n");
    
    // Test Case 4: High-value heavy item vs multiple light items
    printf("Test Case 4: Heavy valuable item vs light items\n");
    Item items4[] = {
        {100, 50, 0.0, 0},  // Ratio: 2.0
        {60, 20, 0.0, 1},   // Ratio: 3.0
        {40, 10, 0.0, 2},   // Ratio: 4.0
        {20, 5, 0.0, 3}     // Ratio: 4.0
    };
    int n4 = 4;
    int capacity4 = 40;
    double selected4[4];
    
    printItems(items4, n4);
    printf("Knapsack capacity: %d\n\n", capacity4);
    
    double maxValue4 = fractionalKnapsack(items4, n4, capacity4, selected4);
    printSelection(items4, n4, selected4, maxValue4, capacity4);
    printf("Greedy correctly chooses lighter items with better ratios\n\n");
    
    // Test Case 5: Edge case - zero capacity
    printf("Test Case 5: Zero capacity knapsack\n");
    Item items5[] = {
        {10, 1, 0.0, 0},
        {20, 2, 0.0, 1}
    };
    int n5 = 2;
    int capacity5 = 0;
    double selected5[2];
    
    printItems(items5, n5);
    printf("Knapsack capacity: %d\n\n", capacity5);
    
    double maxValue5 = fractionalKnapsack(items5, n5, capacity5, selected5);
    printSelection(items5, n5, selected5, maxValue5, capacity5);
    printf("Expected: No items selected due to zero capacity\n");
    
    return 0;
}